new doc: player-compatibility-testing (#125)
RetroAchievements/docs: docs/guidelines/content/player-compatibility-testing.md

---
title: Player Compatibility Testing
description: Describes the process for players to conduct compatibility testing for additional hashes to primary achievement sets on RetroAchievements.
---

# Player Compatibility Testing
The biggest barrier to adding additional hash support to sets is the time required to thoroughly test existing achievement sets against new hashes. This document explains the process by which players may request approval to conduct hash compatibility testing.

## Process Overview

> If approved for a compatibility test, the requested hash will be linked with a **COMPATIBILITY TEST - NOT FOR PLAYER USE** label. The compatibility tester will then be required to master or complete the set within a specified, reasonable time frame. The tester will make save states prior to as many achievements as possible. The tester will report issues directly to the set maintainer via site message and shall not open tickets if using the test hash. If the tester is able to validate that the set is fully compatible through mastery/completion, the hash will be renamed to its appropriate title and its patch will be added to RAPatches. The hash will be considered supported at this point.

Reviewer: ```suggestion
If approved for a compatibility test, the requested hash will be linked with a **COMPATIBILITY TEST - NOT FOR PLAYER USE** label. The compatibility tester will then be required to master or complete the set within a specified, reasonable time frame. The tester should create save states before as many achievements as possible. The tester will report issues directly to the set maintainer via site message and shall not open tickets if using the test hash. If the tester is able to validate that the set is fully compatible through mastery or completion, the hash will be renamed to its appropriate title and its patch will be added to RAPatches. The hash will be considered supported at this point.
```

Author: Manually added as suggested.

> If approved for a compatibility test, the requested hash will be linked with a **COMPATIBILITY TEST - NOT FOR PLAYER USE** label. The compatibility tester will then be required to master or complete the set within a specified, reasonable time frame. The tester will make save states prior to as many achievements as possible. The tester will report issues directly to the set maintainer via site message and shall not open tickets if using the test hash. If the tester is able to validate that the set is fully compatible through mastery/completion, the hash will be renamed to its appropriate title and its patch will be added to RAPatches. The hash will be considered supported at this point.

Reviewer: ```suggestion
If approved for a compatibility test, the requested hash will be linked with a **COMPATIBILITY TEST - NOT FOR PLAYER USE** label. The compatibility tester will then be required to master or complete the set within a specified, reasonable time frame. The tester will make save states prior to as many achievements as possible. The tester will report issues directly to the set maintainer via site message and shall not open tickets if using the test hash. If the tester validates that the set is fully compatible by mastering or completing it, the hash will be renamed. Its patch will then be added to RAPatches. The hash will be considered supported at this point.
```

Author: Manually added as suggested.

## Benefits
Because the test hashes will be linked to the set during testing, testers will keep any achievements they unlock. They will also earn beaten, completion and mastery credit as appropriate.

# Request, Approval, and Testing Process

> Players interested in conducting new hash testing will contact the set maintainer via site direct message and request to conduct compatibility testing. A set maintainer is any active developer who is an author on a set. If all set authors are inactive, [QATeam](https://retroachievements.org/user/QATeam) is the set maintainer.

Reviewer: ```suggestion
Players interested in conducting new hash testing should contact the set maintainer via site direct message to request to conduct compatibility testing. A set maintainer is any active developer who is an author on a set. If all set authors are inactive, [QATeam](https://retroachievements.org/user/QATeam) is the set maintainer.
```

Author: Manually added as suggested.

## Restrictions and Limitations
The following restrictions and limitations apply:

- Set maintainers may deny any request at their discretion

> - Set maintainer may terminate a compatibility test at their discretion

Reviewer: ```suggestion
- The set maintainer may terminate a compatibility test at their discretion
```

Author: Manually added as suggested along with a few other instances.

- Testers must use the emulator and core specified by the set maintainer to conduct all testing
- Depending on how long a set is likely to take to master, a reasonable time limit will be determined by the set maintainer for the tester to complete the set

> - Test hashes that are not actively being tested will be unlinked and associated tests terminated

Reviewer: ```suggestion
- Test hashes that are not actively being tested will be unlinked, and the associated tests will be ended
```

- Only one tester will be approved per test hash

> - Set maintainer may set additional requirements at their discretion

Reviewer: ```suggestion
- The set maintainer may set additional requirements at their discretion
```

Author: Manually added as suggested along with a few other instances.

### Step 1 - The Player's Request

> Players must include the following in their initial compatibility test request message:

Reviewer: ```suggestion
Players must include the following information in their first message to request a compatibility test:
```

Author: Manually added as suggested.

- A brief explanation of why they think a particular hash should be added to the set
- A link to the patch's source
- Explicit commitment that the player intends to master/complete the set

### Step 2 - The Set Maintainer's Conditions
If a set maintainer is interested in supporting the test, they should respond to the player with the following information:
 
- Confirmation that they are willing to permit the compatibility test
- A date by which the test must be completed. This may be adjusted as necessary by the set maintainer at their discretion
- Any other restrictions or requirements not otherwise explained in this doc

### Step 3 - The Player's Agreement
If a player agrees to the terms set forth by the set maintainer, they must respond and acknowledge such.

### Step 4 - The Set Maintainer Links the Test Hash
The set maintainer will link the hash and label it **COMPATIBILITY TEST - NOT FOR PLAYER USE** and inform the tester.

### Step 5 - Testing the Hash
The player will do the following during the compatibility test:

- Make the compatibility testing their #1 priority on RetroAchievements by dedicating the majority of their play time to the test

> - Make a save state prior to each achievement to the maximum extent possible

Reviewer: ```suggestion
- Create a save state before each achievement as much as possible
```

Author: Manually added as suggested.

> - Attempt to earn every achievement in the set that is required by the set maintainer. Sets may require full mastery/completion at the set maintainer's discretion.

Reviewer: ```suggestion
- Attempt to earn every achievement in the set as required by the set maintainer. Sets may require full mastery or completion at the set maintainer's discretion.
```

Author: Manually added as suggested.

- Immediately inform the set maintainer via site message when an incompatible achievement is discovered. The player will be responsible for providing the set maintainer whatever save states they request in order to modify achievement logic necessary for compatibility.

> - The set maintainer may direct additional tests after adjusting logic at their discretion. Testers are required to comply in order for the test to continue

Reviewer: ```suggestion
- The set maintainer may require additional tests after updating the logic. Testers are required to comply in order for the test to continue
```

Author: Manually added as suggested.

### Step 6 - Test Completion

> When the tester masters/completes the set, the test will be considered concluded. At this time, the set maintainer will make the decision whether or not to officially support the test hash. If supported, the set maintainer must submit the patch via the RAPatches forum on the Discord server or by opening a pull request on the RAPatches GitHub repository. Once the patch is added to the repository, the hash description, label, and patch URL must be updated.

Reviewer: ```suggestion
When the tester masters/completes the set, the test will be considered concluded. At this time, the set maintainer will decide whether to officially support the test hash. If supported, the set maintainer must submit the patch via the RAPatches forum on the Discord server or by opening a pull request on the RAPatches GitHub repository. Once the patch is added to the repository, the hash description, label, and patch URL must be updated.
```

Author: Manually added as suggested.
